handle InvalidCodePointException and add a testcase

JoJoBizarreAdventure opened the pull request:
Inherit changes in https://github.com/trinodb/trino/pull/7731/commits
Merge all changes and testcase in one file
handling this issue https://github.com/trinodb/trino/issues/7597

fixes https://github.com/trinodb/trino/issues/7597

skrzypo987 commented on core/trino-main/src/test/java/io/trino/type/TestCharacterStringCasts.java:
```diff
@@ -67,6 +67,12 @@ public void testVarcharToCharSaturatedFloorCast()
         String maxCodePoint = new String(Character.toChars(Character.MAX_CODE_POINT));
         String codePointBeforeSpace = new String(Character.toChars(' ' - 1));
 
+        // Keep decreasing until valid again
```
That comment seems unnecessary. The test is pretty straightforward

skrzypo987 commented on core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java:
```diff
@@ -111,7 +111,17 @@ public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long
+         * Since the codePoint is originally valid, so the only case will be 0XE00 - 1
+         * So we let it go through this range and become 0xD7FF
+         */
+        if (lastCodePoint == 0xDFFF) {
```
Use `Character.MIN_SURROGATE` and `Character.MAX_SURROGATE` constants

skrzypo987 commented on core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java:
```diff
@@ -111,7 +111,17 @@ public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long
-        codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
+        int lastCodePoint = codePoints.get(codePoints.size() - 1) - 1;
+        /*
+         * UTF-8 reserve codepoint from 0xD800 to 0xDFFF for encoding UTF-16
```
codepoint -> codepoints

JoJoBizarreAdventure replied:
ok, I will fix it later

skrzypo987 reviewed (commented):
Looks good.
Please rewrite commit message to something like "Handle InvalidCodePointException in CharacterStringCasts". No need to specify that tests are added.

sopel39 commented:
@JoJoBizarreAdventure please sign cla form

findepi commented:
LGTM overall % @skrzypo987 's comments.

Would be good to update the commit message. Maybe
```
Fix saturated char/varchar cast for U&DFFF
```

JoJoBizarreAdventure commented:
> LGTM overall % @skrzypo987 's comments.
> 
> Would be good to update the commit message. Maybe
> 
> ```
> Fix saturated char/varchar cast for U&DFFF
> ```

Ok, I will squash the commit and update the commit message later

JoJoBizarreAdventure commented:
I have updated the commit and changed the commit message.
Please check it. 
And I wonder if I should change the hex number in testcase to **Character.MIN_SURROGATE** too.

skrzypo987 commented:
> And I wonder if I should change the hex number in testcase to **Character.MIN_SURROGATE** too.

I'd leave the test case as it is.

skrzypo987 reviewed (approved):
Please fill out the CLA, and send a photo/scan to cla@trino.io. Without that we are unable to merge the PR.

JoJoBizarreAdventure commented:
> Please fill out the CLA, and send a photo/scan to [cla@trino.io](mailto:cla@trino.io). Without that we are unable to merge the PR.

I have sent one these days and waiting for reply now.

JoJoBizarreAdventure commented:
I have sent a CLA request but the recent commit doesn't add me to the contributors. Shall I send it again? In usual case how long will the response take?

sopel39 commented:
@cla-bot check

martint commented:
@JoJoBizarreAdventure, can you try re-sending your CLA? We never received it.

JoJoBizarreAdventure commented:
> @JoJoBizarreAdventure, can you try re-sending your CLA? We never received it.

Ok,I will re-sending one with gmail, maybe changing the email will work.

JoJoBizarreAdventure commented:
> @JoJoBizarreAdventure, can you try re-sending your CLA? We never received it.

I have re-sent one, please check.
![image](https://user-images.githubusercontent.com/64888038/117241776-6dfb1200-ae66-11eb-9094-b2bce8502d7b.png)

sopel39 commented:
@cla-bot check

JoJoBizarreAdventure commented:
@cla-bot check

JoJoBizarreAdventure commented:
I have cla-signed now, please check.

martint commented on core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java:
```diff
@@ -111,7 +111,17 @@ public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long
-        codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
+        int lastCodePoint = codePoints.get(codePoints.size() - 1) - 1;
+        /*
+         * UTF-8 reserve codepoints from 0xD800 to 0xDFFF for encoding UTF-16
```
That's actually not standard UTF-8. There's a variant that does this called CESU-8 (see https://en.wikipedia.org/wiki/CESU-8). Note that Trino is expected to deal with proper UTF-8, so it's not clear to me that we actually want this.

skrzypo987 replied:
That is the whole point. Now there it is possible to get an incorrect UTF8 string from the saturated floor cast of the correct one.
`saturatedFllorCast(2, "0xE000") == "0xDFFF..."`

martint replied:
That would not be an incorrect UTF-8 string. It would be an incorrect CESU-8 string

skrzypo987 replied:
Yet it will produce an InvalidCodePointException. Floor cast should not do that.

martint replied:
That's a different problem. Where is that error being thrown? We should fix that issue, instead.

skrzypo987 replied:
Airlift UTF-8 utils will throw exception if the string contains characters like that.

martint replied:
I misunderstood the purpose of this patch. The proposed fix is correct

colebow commented:
@martint - do we want a release note for this?
